Add array API support to `det_curve`

jaffourt opened the pull request:
#### Reference Issues/PRs
Towards #26024

#### What does this implement/fix? Explain your changes.
Add array API support to `det_curve`

#### Any other comments?

jaffourt committed:
add array api support for `det_curve`

jaffourt committed:
add towncrier rst

jaffourt committed:
ensure dtypes match

jaffourt committed:
use `concat` not `concatenate`

jaffourt committed:
fix dtype

jaffourt committed:
fix

jaffourt committed:
typo

jaffourt commented:
CC: @OmarManzoor @virchan @betatim @lucyleeow In case any of you have time to review :)

lucyleeow reviewed (commented):
Thanks for your contribution.

Could you please disclose if you used AI to help with this PR? Thanks

(please see: https://scikit-learn.org/dev/developers/contributing.html#automated-contributions-policy)

jaffourt committed:
Apply suggestions from code review

Co-authored-by: Lucy Liu <jliu176@gmail.com>

jaffourt committed:
remove specific numpy test use comon

jaffourt committed:
handle mismatched dtypes with xp.inf

jaffourt commented:
@lucyleeow Thank you for the review! I think I addressed all your comments.

> Could you please disclose if you used AI to help with this PR? Thanks

Only to create the towncrier thing, the rest was me :)

lucyleeow commented on sklearn/utils/validation.py:
```diff
@@ -2612,15 +2612,18 @@ def _check_pos_label_consistency(pos_label, y_true):
+                )
+                or xp.all(classes == xp.asarray([0], dtype=cls_dtype, device=device))
+                or xp.all(classes == xp.asarray([-1], dtype=cls_dtype, device=device))
+                or xp.all(classes == xp.asarray([1], dtype=cls_dtype, device=device))
```
We always assume `y_true` to be int; sklearn estimators would not output labels as floats. We also often use `np.unique` on true labels, which isn't ideal with floats due to floating point precision.

Others can comment more but changing to allow labels to be floats would need to be discussed and would be outside the scope of this PR.

jaffourt committed:
Revert "ensure dtypes match"

This reverts commit 54f4b533d1cfa0bb9ae0205d776711ea769c6c3b.

OmarManzoor reviewed (approved):
LGTM. Thanks @jaffourt

OmarManzoor commented on sklearn/metrics/_ranking.py:
```diff
@@ -372,15 +372,18 @@ def det_curve(
+    tps = xp.concat((xp.asarray([0.0], device=device), tps))
+    fps = xp.concat((xp.asarray([0.0], device=device), fps))
+    thresholds = xp.astype(thresholds, _max_precision_float_dtype(xp, device))
+    thresholds = xp.concat((xp.asarray([xp.inf], device=device), thresholds))
```
Maybe we can avoid changing `thresholds` and instead adjust the `dtype` on the temp array that we are creating
```suggestion
    thresholds = xp.concat((xp.asarray([xp.inf], dtype=thresholds.dtype, device=device), thresholds))
```
Or do we need to adjust the `dtype` of `thresholds` specifically

lucyleeow reviewed (approved):
LGTM thanks!
